File: 2023/24.py
from utils import tuple_sum, tuple_diff


def intersect_2d(h1, h2, r1, r2):
    p1, p2 = h1["p"][:2], tuple_sum(h1["p"][:2], h1["v"][:2])
    p3, p4 = h2["p"][:2], tuple_sum(h2["p"][:2], h2["v"][:2])
    a1, a2 = -h1["v"][0], -h1["v"][1]
    b1, b2 = -h2["v"][0], -h2["v"][1]
    det = a1 * b2 - a2 * b1
    if det == 0:
        return False
    c1, c2 = p1[0]*p2[1] - p1[1]*p2[0], p3[0]*p4[1] - p3[1]*p4[0]
    x = (c1 * b1 - a1 * c2) / det
    y = (c1 * b2 - a2 * c2) / det
    if tuple_diff(p1, (x, y))[0] * a1 < 0 or tuple_diff(p3, (x, y))[0] * b1 < 0:
        return False
    return r1 <= x <= r2 and r1 <= y <= r2


def part_one(hailstones):
    crossed = 0
    for i, h1 in enumerate(hailstones):
        for h2 in hailstones[i+1:]:
            crossed += intersect_2d(h1, h2, 200_000_000_000_000, 400_000_000_000_000)
    return crossed


def part_two(hailstones):
    return 0


def main(res):
    with open("input.txt", "r") as f:
        lines = [line.strip().split(" @ ") for line in f.readlines()]
        hailstones = [{
            "p": list(map(int, line[0].split(","))),
            "v": list(map(int, line[1].split(","))),
        } for line in lines]
    return res(hailstones)


if __name__ == "__main__":
    print(main(part_one))
    # print(main(part_two))

File: 2024/24.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use itertools::Itertools;

fn part_one(wires: &mut HashMap<String, bool>, gates: &Vec<Vec<String>>) -> String {
    let mut done = false;
    while !done {
        done = true;
        for gate in gates {
            if !wires.contains_key(&gate[0]) || !wires.contains_key(&gate[2]) {
                done = false;
                continue;
            }
            let a = wires.get(&gate[0]).unwrap().clone();
            let b = wires.get(&gate[2]).unwrap().clone();
            wires.insert(gate[4].clone(), match gate[1].as_str() {
                "AND" => a & b,
                "OR" => a | b,
                _ => a ^ b,
            });
        }
    }
    wires.iter()
        .filter(|(k, _)| k.starts_with("z"))
        .sorted_by(|a, b| a.0.cmp(&b.0))
        .enumerate()
        .map(|(i, (_, &v))| if v { 2_u64.pow(i as u32) } else { 0 })
        .sum::<u64>()
        .to_string()
}

fn part_two(_wires: &mut HashMap<String, bool>, _gates: &Vec<Vec<String>>) -> String {
    "".to_string()
}

fn main_fn(res: &dyn Fn(&mut HashMap<String, bool>, &Vec<Vec<String>>) -> String) -> String {
    let reader: BufReader<File> = BufReader::new(File::open("input").unwrap());
    let mut wires: HashMap<String, bool> = HashMap::new();
    let mut gates: Vec<Vec<String>> = Vec::new();
    let mut inputs = true;
    for line in reader.lines().flatten() {
        if line.is_empty() {
            inputs = false;
            continue;
        }
        if inputs {
            let (key, value) = line.split(": ").map(|s| s.to_string()).collect_tuple().unwrap();
            wires.insert(key, value.parse::<u8>().unwrap() != 0);
        } else {
            gates.push(line.split_whitespace().map(|s| s.to_string()).collect());
        }
    }
    res(&mut wires, &gates)
}

fn main() {
    println!("{}", main_fn(&part_one));
    println!("{}", main_fn(&part_two));
}
